### PR TITLE
Improvements for parameterize

### PR DESCRIPTION
**Pull request by stefdoerr**
1. Increased tolerance of charge fitting to something realistic
2. Increased the upper lower bounds of fitting
3. Hacked around in the minimizer to attempt multiple minimizations before giving up

Linked to: https://github.com/Acellera/htmd-data/pull/33

### CODE REVIEW

File: htmd/qm/custom.py [hunk 1]
```diff
@@ -130,8 +130,34 @@ def goalFunc(x):
             grad = -forces.reshape(-1)
             return energy, grad
 
-        res = minimize(goalFunc, coords.reshape(-1), method='L-BFGS-B', jac=True, options={'ftol': 0, 'gtol': 0.05})
-        endcoords = res.x.reshape((natoms, 3)).copy()
+        gtol = 0.1  # kcal/mol/A
```
Comment (raimis) [inline htmd/qm/custom.py]: Just keep the code human-readable: `gtol` --> 'force_tolernace'

File: htmd/qm/custom.py [hunk 2]
```diff
@@ -130,8 +130,34 @@ def goalFunc(x):
...
+        maxattempts = 50
```
Comment (raimis) [inline htmd/qm/custom.py]: `maxattempts` --> `max_attempts`

File: htmd/qm/custom.py [hunk 3]
```diff
@@ -130,8 +130,34 @@ def goalFunc(x):
...
+        bestres = None
```
Comment (raimis) [inline htmd/qm/custom.py]: `bestres` --> `best_result`

File: htmd/qm/custom.py [hunk 4]
```diff
@@ -130,8 +130,34 @@ def goalFunc(x):
...
+        bestforces = np.inf
```
Comment (raimis) [inline htmd/qm/custom.py]: `bestforces` --> `best_force`

File: htmd/qm/custom.py [hunk 5]
```diff
@@ -130,8 +130,34 @@ def goalFunc(x):
...
+        for i in range(maxattempts):
+            res = minimize(goalFunc, coords.reshape(-1), method='L-BFGS-B', jac=True, options={'ftol': 0, 'gtol': gtol})
```
Comment (raimis) [inline htmd/qm/custom.py]: `res` --> `result`

File: htmd/qm/custom.py [hunk 6]
```diff
@@ -130,8 +130,34 @@ def goalFunc(x):
...
+            maxforcecomp = np.abs(res.jac).max()
```
Comment (raimis) [inline htmd/qm/custom.py]: `maxforcecomp` --> `max_force`

File: htmd/qm/custom.py [hunk 7]
```diff
@@ -130,8 +130,34 @@ def goalFunc(x):
...
+
+            if maxforcecomp < bestforces:
+                bestforces = maxforcecomp
+                bestres = res
+
+            if maxforcecomp > gtol:
```
Comment (raimis) [inline htmd/qm/custom.py]: One attempt is one minimization. Why do you need the second one?

File: htmd/qm/custom.py [hunk 8]
```diff
@@ -130,8 +130,34 @@ def goalFunc(x):
...
+                bestforces = maxforcecomp
+                bestres = res
+
+            if maxforcecomp <= gtol:
+                break
+
+        if bestforces > gtol:
+            logger.warning('Did not manage to minimize structure to the desired force tolerance. Best minimized structure had a max force component {:.2f} kcal/mol/A. Threshold is {}'.format(bestforces, gtol))
+
+        endcoords = bestres.x.reshape((natoms, 3)).copy()
```
Comment (raimis) [inline htmd/qm/custom.py]: `endcoords` --> `minimized_coords`

**Comment by stefdoerr** [comment]
Seriously. I'll do the changes but I don't think I'll bother again after this PR. We don't have a coding style for HTMD on variable naming and we can fight over it for ages. If anything the "official" is to use camel case and not underscores.

**Comment by stefdoerr** [comment]
Otherwise let's sit down and decide on a naming style.

File: htmd/qm/custom.py (hunk 7 above)
Comment (stefdoerr) [reply htmd/qm/custom.py]: Resets the hessian

**Comment by raimis** [comment]
Sorry @stefdoerr , I just following the code around.

File: htmd/qm/custom.py (hunk 7 above)
Comment (raimis) [reply htmd/qm/custom.py]: The Hessian is reset with each minimization. Still I don't get the point, why do you two per attempt, you could just loop 100 attempts instead.

**Comment by stefdoerr** [comment]
Renamed everything

File: htmd/qm/custom.py (hunk 7 above)
Comment (stefdoerr) [reply htmd/qm/custom.py]: It continues where the previous one quit. See what I pass it as starting
coords. It's the result of the above minimization

Comment (stefdoerr) [reply htmd/qm/custom.py]: I had cases where restarting from the previous "near minimum" finished the minimization fine.

**Comment by raimis** [comment]
@stefdoerr it seems your are missing a PR for `htmd-data`.

**Comment by raimis** [comment]
The test on my machine fails. The number are bit different, so it is an instability.

Let's increase the tolerances:
```diff
diff --git a/htmd/parameterization/test.py b/htmd/parameterization/test.py
index 6c2a6ff16..c1619477d 100644
--- a/htmd/parameterization/test.py
+++ b/htmd/parameterization/test.py
@@ -259,7 +259,7 @@ class TestParameterize(unittest.TestCase):
         resDir = os.path.join(self.testDir, 'ethene_dihed_fix_restart')
         shutil.copytree(os.path.join(refDir, 'dihedral-single-point'), os.path.join(resDir, 'dihedral-single-point'))
         self._run(refDir, resDir, 'parameterize input.mol2 --charge-type Gasteiger --min-type None --scan-type None')
-        self._test(refDir, resDir)
+        self._test(refDir, resDir, energyTermRelTol=5e-6, dihedralForceConstAbsTol=5e-6)
 
     @unittest.skipUnless(os.environ.get('HTMD_VERYLONGTESTS') == 'yes', 'Too long')
     def test_glycol_dihed_fix(self):
```

**Comment by raimis** [comment]
Victory!!! @stefdoerr any comment before merging?

**Comment by stefdoerr** [comment]
Nap

**Pull request merged**